src/TelemetryConsumers/Orleans.TelemetryConsumers.Linux/LinuxEnvironmentStatistics.cs

```diff
@@ -0,0 +1,276 @@
+#define LOG_MEMORY_PERF_COUNTERS
+
+using System;
+using System.IO;
+using System.Linq;
+using System.Runtime.InteropServices;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using Orleans.Runtime;
+
+namespace Orleans.Statistics
+{
+    internal class LinuxEnvironmentStatistics : IHostEnvironmentStatistics, ILifecycleParticipant<ISiloLifecycle>, ILifecycleParticipant<IClusterClientLifecycle>, ILifecycleObserver, IDisposable
+    {
+        private readonly ILogger _logger;
+
+        private const float KB = 1024f;
+
+        /// <inheritdoc />
+        public long? TotalPhysicalMemory { get; private set; }
+
+        /// <inheritdoc />
+        public float? CpuUsage { get; private set; }
+
+        /// <inheritdoc />
+        public long? AvailableMemory { get; private set; }
+
+        /// <inheritdoc />
+        public long MemoryUsage => GC.GetTotalMemory(false);
+
+        private readonly TimeSpan MONITOR_PERIOD = TimeSpan.FromSeconds(5);
+
+        private CancellationTokenSource _cts;
+        private Task _monitorTask;
+
+        private static readonly string StartErrMsg_Win = $"Tried to start {nameof(LinuxEnvironmentStatistics)}, but detected OS Windows. " +
+            "Call 'UsePerfCounterEnvironmentStatistics' instead for implementation based on Windows Performance Counters";
+
+        private static readonly string StartErrMsg_Unknown = $"Tried to start {nameof(LinuxEnvironmentStatistics)}, " +
+            $"but detected OS is not Linux: '{RuntimeInformation.OSDescription}'";
+
+        public LinuxEnvironmentStatistics(ILoggerFactory loggerFactory)
+        {
+            _logger = loggerFactory.CreateLogger<LinuxEnvironmentStatistics>();
+        }
+
+        public void Dispose()
+        {
+            _cts?.Dispose();
+            _monitorTask?.Dispose();
+        }
+
+        public void Participate(ISiloLifecycle lifecycle)
+        {
+            lifecycle.Subscribe(ServiceLifecycleStage.RuntimeInitialize, this);
+        }
+
+        public void Participate(IClusterClientLifecycle lifecycle)
+        {
+            lifecycle.Subscribe(ServiceLifecycleStage.RuntimeInitialize, this);
+        }
+
+        public async Task OnStart(CancellationToken ct)
+        {
+            var isLinux = RuntimeInformation.IsOSPlatform(OSPlatform.Linux);
+
+            if (!isLinux)
+            {
+                if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+                    _logger.LogWarning(StartErrMsg_Win);
+                else
+                    _logger.LogWarning(StartErrMsg_Unknown);
+                return;
+            }
+
+            _logger.LogTrace($"Starting {nameof(LinuxEnvironmentStatistics)}");
+
+            _cts = new CancellationTokenSource();
+            ct.Register(() => _cts.Cancel());
+
+            _monitorTask = await Task.Factory.StartNew(
```

Comment:
Not 100% sure, but woudln't it be better to simply use a `Timer` for that task instead of a look with a `Wait`?

Reply:
`Wait` uses `Timer` underneath AFAIK, and personally think just the loop looks simpler. Will change if you want